docs/proposal/aws-resource-handling.md

```diff
@@ -0,0 +1,200 @@
+# AWS Resource Handling
+
+## Problem
+
+Since the AWS APIs do not provide consistent support for tagging resources on creation or idempotent operations, we need to ensure that we can provide as much of an external guarantee to users as we can. Otherwise we risk creating and orphaning resources on users leading to extraneous costs and the potential to exhaust resource quotas prematurely.
+
+## Existing solutions and drawbacks for our use
+
+### kops
+
+kops relies heavily on tagging of resources. This provides the benefit of recreating the state of the cluster resources by querying with filters. However, in many cases tagging requires a second API call and there are some resources that are not able to be tagged. If we succeed in creating a resource but fail to tag the resource, then we risk the chance of orphaning that resource for the user.
+
+### Kubicorn
+
+In contrast to the kops approach, Kubicorn mainly relies on recording the resource IDs as part of state. However, since we rely on an external API server for recording the state there is still a possibility of creating the resource and failing to record the resource ID, which still exposes the possible risk of orphaning that resource for the user.
+
+## Summary of edge cases for creating an individual resource
+
+1. resource create succeeds, but subsequent tagging fails
+2. resource creates succeeds, but update of cluster/machine object fails
+3. attempting to delete resource fails after an attempt to rollback due to a failure to record the ID of the created resource to the cluster/machine object for resources that do not support tagging on create.
```

Comment:
highlighting this since it's a good idea and maybe hasn't been explicitly stated: if the object store goes away and we cannot save references to the resources we create then we should rollback.

Reply:
This should only be necessary in the case that the operation does not support client tokens or tagging as part of create based on the workflows proposed below.

Reply:
Not sure I prefer using Tags over ResourceIDs by default. We did [rely on tags for resources in the past](https://github.com/scalefactory/build-cloud), but then never sorted out what to do for resources that didn't have them (@jtopper can speak more authoritatively on that one). At some point we need to track ARNs or other resource IDs, and it might make more sense to use that universally? 

Do we know the cases in which the controllers would lose access to the API Server as a result of a change it's made? If it's during cluster turndown, could we "unpivot" the controller instead?

Reply:
The approach we used in build-cloud was similar to that used when building a Puppet provider: we'd choose an attribute that looked the most like a "primary key" for that type of object to track, and if there wasn't an obvious one from the resource, we'd use a name tag. There are some resource types where that wasn't possible because tagging wasn't available for that type; and in some cases the parameter we chose as a "primary key" can't be guaranteed unique by the platform - but we could assume that it was in our own usage. I think tracking resource IDs is probably the more trustworthy approach here, even though it'll take more engineering effort.